### PR TITLE
fix!: Migrate file destination to SDK v2

### PR DESCRIPTION
**Pull request by hermanschaaf**
This change does the following:
 - updates to the new plugin SDK v2 and filetypes v2. There are no user-facing changes for CSV or JSON, but the format of UUID and Datetime fields changes for Parquet. 
 - adds a new `path` parameter to allow for more flexible naming of the output files. The old `directory` parameter is still supported, but will be removed in a future version and for now a warning is logged when it is used. When upgrading to v2.0.0, please consider changing from `directory` to `path` as well.

BEGIN_COMMIT_OVERRIDE
feat: Update to use [Apache Arrow](https://arrow.apache.org/) type system
feat: Introduce `path` parameter with placeholder variables to replace `directory`. Please consider updating your config to use `path` instead of `directory`.

BREAKING-CHANGE: This release introduces an internal change to our type system to use [Apache Arrow](https://arrow.apache.org/). This should not have any visible breaking changes for CSV or JSON output formats, however the Parquet output changes for UUID columns, which now have dashes, and timestamps, which now uses the default Arrow time format (e.g. `2023-01-02 12:23:45`). If you encounter an issue during the upgrade, please submit a [bug report](https://github.com/cloudquery/cloudquery/issues/new/choose).

END_COMMIT_OVERRIDE

Closes https://github.com/cloudquery/cloudquery/issues/10114

### CODE REVIEW

File: plugins/destination/file/go.mod [hunk 1]
```diff
@@ -2,43 +2,37 @@ module github.com/cloudquery/cloudquery/plugins/destination/file
...
 	github.com/minio/c2goasm v0.0.0-20190812172519-36a3d3bbc4f3 // indirect
+	github.com/rs/zerolog v1.29.0
 	github.com/zeebo/xxh3 v1.0.2 // indirect
 	golang.org/x/exp v0.0.0-20230321023759-10a507213a29 // indirect
 	golang.org/x/mod v0.9.0 // indirect
 	golang.org/x/tools v0.7.0 // indirect
 )
 
-replace github.com/apache/arrow/go/v12 => github.com/cloudquery/arrow/go/v12 v12.0.0-20230317130341-c648117570af
+replace github.com/cloudquery/filetypes => ../../../../filetypes
```
Comment (disq) [inline plugins/destination/file/go.mod:24]: replace github.com/cloudquery/filetypes => github.com/cloudquery/filetypes v1.6.3-0.20230417092324-d4eaee234b77

File: plugins/destination/file/go.mod [hunk 2]
```diff
@@ -2,43 +2,37 @@ module github.com/cloudquery/cloudquery/plugins/destination/file
...
+replace github.com/apache/arrow/go/v12 => ../../../../arrow/go
```
Comment (disq) [inline plugins/destination/file/go.mod:25]: replace github.com/apache/arrow/go/v12 => github.com/cloudquery/arrow/go/v12 v12.0.0-20230416141001-e8315b3fff1c

File: plugins/destination/file/go.mod (hunk 1 above)
Comment (disq) [inline plugins/destination/file/go.mod:24]:
```suggestion
replace github.com/apache/arrow/go/v12 => github.com/cloudquery/arrow/go/v12 v12.0.0-20230416141001-e8315b3fff1c
```

File: plugins/destination/file/go.mod (hunk 2 above)
Comment (disq) [inline plugins/destination/file/go.mod:25]:
```suggestion
replace github.com/apache/arrow/go/v12 => github.com/cloudquery/arrow/go/v12 v12.0.0-20230416141001-e8315b3fff1c
```

File: plugins/destination/file/go.mod [hunk 3]
```diff
@@ -2,43 +2,37 @@ module github.com/cloudquery/cloudquery/plugins/destination/file
...
+
+replace github.com/cloudquery/plugin-sdk/v2 => ../../../../plugin-sdk
```
Comment (disq) [inline plugins/destination/file/go.mod:27]:
```suggestion
```

File: plugins/destination/file/go.mod [hunk 4]
```diff
@@ -2,43 +2,37 @@ module github.com/cloudquery/cloudquery/plugins/destination/file
...
-	github.com/google/uuid v1.3.0
-	github.com/rs/zerolog v1.29.0
-)
-
 require (
 	github.com/andybalholm/brotli v1.0.5 // indirect
-	github.com/apache/arrow/go/v12 v12.0.0-20230331222054-7e19111f2f81 // indirect
+	github.com/apache/arrow/go/v12 v12.0.0-20230411095929-e488942cd552
+	github.com/cloudquery/filetypes v1.6.2
+	github.com/cloudquery/plugin-sdk/v2 v2.0.0
```
Comment (disq) [inline plugins/destination/file/go.mod:9]:
```suggestion
	github.com/cloudquery/plugin-sdk/v2 v2.2.0
```

**Comment by hermanschaaf** [comment]
This PR is now almost ready - it works in local testing, but the tests warn us about some memory leaks that I need to track down next.

File: plugins/destination/file/client/read.go [hunk 5]
```diff
@@ -4,20 +4,28 @@ import (
...
+	"github.com/cloudquery/plugin-sdk/v2/schema"
+	"github.com/google/uuid"
 )
 
-func (c *Client) Read(ctx context.Context, table *schema.Table, sourceName string, res chan<- []any) error {
+func (c *Client) Read(ctx context.Context, arrowSchema *arrow.Schema, sourceName string, res chan<- arrow.Record) error {
+	tableName := schema.TableName(arrowSchema)
 	if !c.pluginSpec.NoRotate {
-		return fmt.Errorf("reading is not supported when no_rotate is false. Table: %q; Source: %q", table.Name, sourceName)
+		return fmt.Errorf("reading is not supported when no_rotate is false. Table: %q; Source: %q", tableName, sourceName)
```
Comment (disq) [inline plugins/destination/file/client/read.go:18]:
```suggestion
		return fmt.Errorf("reading is not supported when `no_rotate` is false. Table: %q; Source: %q", tableName, sourceName)
```
More changes along the same lines as a general clean up?

File: plugins/destination/file/client/spec.go [hunk 6]
```diff
@@ -2,25 +2,53 @@ package client
...
+}
 
 func (s *Spec) Validate() error {
-	if s.Directory == "" {
-		return fmt.Errorf("directory is required")
+	if s.Directory == "" && s.Path == "" {
+		return fmt.Errorf("either directory or path must be set")
+	}
+	if s.Directory != "" && s.Path != "" {
+		return fmt.Errorf("only one of directory or path is allowed")
```
Comment (disq) [inline plugins/destination/file/client/spec.go:48]:
```suggestion
		return fmt.Errorf("only one of `directory` or `path` is allowed")
```

File: plugins/destination/file/client/spec.go [hunk 7]
```diff
@@ -2,25 +2,53 @@ package client
...
 	}
 	if s.Format == "" {
 		return fmt.Errorf("format is required")
```
Comment (disq) [inline plugins/destination/file/client/spec.go:51]:
```suggestion
		return fmt.Errorf("`format` is required")
```

File: website/pages/docs/plugins/destinations/file/overview.mdx [hunk 8]
```diff
@@ -10,21 +10,36 @@ import Configuration from "./_configuration.mdx";
...
-This is the (nested) spec used by the CSV destination Plugin.
+This is the (nested) spec used by the file destination Plugin.
 
-- `directory` (string) (required)
+- `path` (string) (required)
 
-  Directory where all CSV files will be written. A CSV file will be created per table.
+  Path template string that determines where files will be written. The path supports the following placeholder variables:
+
+  - `{{TABLE}}` will be replaced with the table name,
```
Comment (disq) [inline website/pages/docs/plugins/destinations/file/overview.mdx:29]: Maybe get rid of the comma suffixes and the final `and` format for ease of editability/maintenance later? Just leave each bullet point "hanging", without a full stop?

Comment (hermanschaaf) [reply website/pages/docs/plugins/destinations/file/overview.mdx]: Yeah, sounds good

**Review by erezrokah — APPROVED** [review]

**Pull request merged**